### DISCUSSION TITLE
How do I redirect for only a few query strings?

### DISCUSSION THREAD

**Discussion by ojj1123** [vercel/next.js#54845] [Help] [answered]
### Summary

# What I want to do
For the path `/category`, the query string `category` accepts only the following values: `'ALL'` `'LOVE'` `'THANKS'` `'CHEERING'` `'CELEBRATE'`
So, in the case of the following path `/category`(no query), `/category?category=ALLL`(wrong query string) or `/category?category=`(empty query string), **I would like to redirect here**: `/category?category=ALL`.

# What I try to do
I edited `redirect` of `next.config.js` options like this:
```js
const nextConfig = {
  async redirects() {
    return [
      {
        source: '/category',
        has: [
          {
            type: 'query',
            key: 'category',
            value: '(\\b(?!(ALL|LOVE|CELEBRATE|THANKS|CHEERING)\\b)\\w+\\b)',
          },
        ],
        destination: '/category?category=ALL',
        permanent: false,
      },
    ];
  },
};

module.exports = nextConfig;
```
But it has a problem: both `/category` and `/category?category=` don't redirect to `/category?category=ALL`.
I want to redirect to `/category?category=ALL` for above routes.

(The path `/category?category=ALLL`(wrong query string) redirect to `/category?category=ALL` what i intend.)

# Questions
1. Above the configuration, Is  `Regex` of `value` property WRONG?
2. I am confused for `redirect` of `next.config.js` options. Can this be solved with the `missing` field instead of the `has` field? I'm confused about the difference between the `missing` field and the `has` field. Please explain!
3. If there is a problem or solution that I didn't think of, please let me know!

# Reference
https://nextjs.org/docs/app/api-reference/next-config-js/redirects

### Additional information

```js
Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 22.5.0: Thu Jun  8 22:22:19 PDT 2023; root:xnu-8796.121.3~7/RELEASE_ARM64_T8103
    Binaries:
      Node: 16.17.0
      npm: 8.19.0
      Yarn: 1.22.19
      pnpm: 8.6.2
    Relevant packages:
      next: 13.4.1
      eslint-config-next: 13.4.1
      react: 18.2.0
      react-dom: 18.2.0
```
```


### Example

_No response_

**Comment by devjiwonchoi** [comment, marked as answer]
Based on my understanding you want the default result to be equal to params `ALL` right?

You can use `middleware` in this case.

Note that this is just one of a way to resolve this:

```ts
// middleware.ts

import { type NextRequest, NextResponse } from 'next/server'

const categories = ['ALL', 'LOVE', 'THANKS', 'CHEERING', 'CELEBRATE']

const hasValidCategory = (category: string) => categories.includes(category)

export default function middleware(req: NextRequest) {
  const category = req.nextUrl.searchParams.get('category')

  if (category) {
    if (!hasValidCategory(category)) {
      // if the category is invalid, we can redirect as /category?category=ALL
      return NextResponse.redirect(
        new URL('/category?category=ALL', req.nextUrl)
      )
    }
    // if the category is valid, we can return early and let the request continue
    return NextResponse.next()
  }
}
```

**Reply by ojj1123** [reply]
@devjiwonchoi  Thanks to you, I resolve my issue! 👍
